### PR TITLE
Add strict i18n template literal rules

### PR DESCRIPTION
**Pull request by KevinMind**
fixes: #232

In the process of investigating https://github.com/mozilla/addons-server/issues/9112

I found that our eslint rules for i18n allow for two edge cases that we should probably forbid or fix.

## 1. non literal strings as arguments.

Passing dynamic content to i18n methods that should return translations make the inputs NOT unpredictable  and therefore extraction cannot generate valid locale entries. You can see examples of invalid reference passing in the tests of this PR but examples are 

- window.foo
- foo()
- this.props.foo

These values could be anything and therefore cannot be tracked in a locale. The reason for this rule is the same as why you would have a rule for no interpolated values.

## 2. Template literals as arguments

Similar to the first rule, this is simply a more strict version of an existing rule. We already don't allow interpolated value expressions in template literals, as well as not allowing tagged template literals.

This rule goes one step further and fixes static template literals to normal strings. There are only 2 reasons to use template literals:

- passing dyanmic expressions
- adding explicit spacing and line breaks.

We already forbid the first case and the second case is artificially removed from our code during extraction making the extraction process more complex than it needs to be. 

By solving this at the linting level we can prevent any template literals from entering the equation at all.

These rules are thoroughly tested against all methods of i18n using our existing configuration for identifying positional arguments in i18n methods we want to protect. This PR introduces 2 utility meethods that allow you to generate test cases against all methods and positions from a static array of predefined inputs, outputs, and errors.

These could be used for other or the existing i18n rule tests, you know, if you want.

### CODE REVIEW

**Comment by KevinMind** [comment]
This [PR](https://github.com/mozilla/addons-frontend/pull/12780) used the eslint rule to autofix in the [first](https://github.com/mozilla/addons-frontend/pull/12780/commits/6609597ec62e89e43e4b9c4dd499ad804e08bb98) commit, and manually fix in the [second](https://github.com/mozilla/addons-frontend/pull/12780/commits/d992090697a3aefa7a1f6b35e237f52a22db1393).

File: lib/rules/i18n-no-reference.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+function extractI18nMethodAndArgs(node, methodName) {
+  let result = [];
+
+  if (node.type === 'CallExpression') {
+    const { callee } = node;
+    if (callee.type === 'MemberExpression') {
+      const { object, property } = callee;
+      if (object.name === 'i18n') {
+        if (!methodName || (methodName !== property.name)) {
```
Comment (KevinMind) [inline lib/rules/i18n-no-reference.js:19]: I think that second check is wrong it should be methodName === property.name. I never actually pass a second arg so probably could just remove it... 😆

Comment (KevinMind) [reply lib/rules/i18n-no-reference.js]: Fixed.

File: lib/rules/i18n-no-reference.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        'invalid argument type "{{ arg }}" passed to i18n.sprintf(). Only static values or calls to i18n methods are allowed',
+    },
+  },
+
+  create(context) {
+    return {
+      CallExpression(node) {
+        const [methodName, methodArgs] = extractI18nMethodAndArgs(node);
+
+        if (!methodName || !methodArgs) return;
```
Comment (willdurand) [inline lib/rules/i18n-no-reference.js:51]: (nit)

```suggestion
        if (!methodName || !methodArgs) {
          return;
        }
```

File: lib/rules/i18n-no-reference.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        if (methodName.includes('ngettext')) return;
```
Comment (willdurand) [inline lib/rules/i18n-no-reference.js:53]:
```suggestion
        if (methodName.includes('ngettext')) {
          return;
        }
```

Comment (willdurand) [reply lib/rules/i18n-no-reference.js:53]: I suppose you could merge both `if` above.

File: lib/rules/i18n-no-reference.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+            if (!argMethod) {
+              context.report({
+                node,
+                messageId: 'noDynamicValueForSprintf',
+                data: {
+                  arg: firstArg.type,
+                  method: methodName,
+                },
+              });
```
Comment (willdurand) [inline lib/rules/i18n-no-reference.js:69]: shouldn't we early return here?

File: lib/rules/i18n-no-reference.js [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            }
+          }
+        }
+
+        let positions = I18N_METHODS[methodName] || [];
```
Comment (willdurand) [inline lib/rules/i18n-no-reference.js:74]: (nit)

```suggestion
        const positions = I18N_METHODS[methodName] || [];
```

File: lib/rules/i18n-no-template-literal.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      CallExpression(node) {
+        const { callee, arguments: args } = node;
+
+        if (!callee.type === 'MemberExpression' || !args.length) {
+          return;
+        }
+
+        const { object, property } = node.callee;
+
+        if (!object || !object.name || object.name !== 'i18n') {
```
Comment (willdurand) [inline lib/rules/i18n-no-template-literal.js:32]: nit: I suppose we could use the optional chaining operator these days:

```suggestion
        if (object?.name !== 'i18n') {
```

File: lib/rules/i18n-no-template-literal.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+          return;
+        }
+
+        let positions = I18N_METHODS[property.name] || [];
```
Comment (willdurand) [inline lib/rules/i18n-no-template-literal.js:36]:
```suggestion
        const positions = I18N_METHODS[property.name] || [];
```

File: lib/rules/i18n-no-template-literal.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+        positions.forEach((position) => {
+          const arg = args[position];
+
+          function toStaticString(quasis) {
```
Comment (willdurand) [inline lib/rules/i18n-no-template-literal.js:41]: any reason to have this function defined here vs. at the top of the file?

File: lib/rules/i18n-no-template-literal.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            const reportObject = {
+              node,
+              messageId: 'noStaticTemplate',
+              data: {
+                method: property.name,
+              }
+            };
+
+            if (!arg.expressions || arg.expressions.length === 0) {
+              reportObject.fix = function templateToString(fixer) {
```
Comment (willdurand) [inline lib/rules/i18n-no-template-literal.js:65]: (nit)

```suggestion
              reportObject.fix = (fixer) => {
```

File: tests/utils.js [hunk 10]
```diff
@@ -0,0 +1,60 @@
+function mapPositionsToArgs(replacement, positions, staticArg) {
```
Comment (willdurand) [inline tests/utils.js:1]: can we use arrow functions here (in this file) too?

Comment (willdurand) [reply tests/utils.js:1]: also, is it assumed that we have test coverage via the added test files for the rules?

File: lib/rules/i18n-no-reference.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+
+module.exports = {
+  meta: {
+    docs: {
+      category: 'Possible Errors',
+      description:
+        'Ensure predictable static values are passed as i18n method arguments',
+      recommended: false,
```
Comment (willdurand) [inline lib/rules/i18n-no-reference.js:34]: If you mark the new rules as `recommended`, they will be automatically used in addons-frontend: https://github.com/mozilla/addons-frontend/blob/8687a2f3d6b4a0366f4b89ec51e160d0a35ffdd0/.eslintrc#L4

Comment (willdurand) [reply lib/rules/i18n-no-reference.js:34]: You could also add a link to the github issue that led to the creation of this rule using: `githubIssue: '<link>'` (which will be used in the generated docs).

File: lib/rules/i18n-no-reference.js [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    },
+    fixable: null, // Indicates that some problems reported by this rule are automatically fixable
+    schema: [],
+    messages: {
+      noDynamicValue:
+        'invalid argument type "{{ arg }} passed to i18n.{{ method }}()".',
```
Comment (willdurand) [inline lib/rules/i18n-no-reference.js:40]: so that would print something like this, right?

```
invalid argument type "window passed to i18n.gettext()".
```

I am wondering if we shouldn't quote the argument only, like for `noDynamicValudForSprintf`:

```
invalid argument type "window" passed to i18n.gettext().
```

maybe we should get rid of the "type" in the middle too? so that it would say that there is an invalid argument passed into the i18n method call?

**Comment by willdurand** [comment]
in case that helps, here is the current code coverage of the PR:

```
-------------------------------------------|---------|----------|---------|---------|-------------------
File                                       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-------------------------------------------|---------|----------|---------|---------|-------------------
All files                                  |   94.49 |    90.37 |     100 |   95.21 |
 lib                                       |     100 |      100 |     100 |     100 |
  i18n.js                                  |     100 |      100 |     100 |     100 |
  index.js                                 |     100 |      100 |     100 |     100 |
 lib/rules                                 |   94.17 |    90.16 |     100 |   94.93 |
  dangerously-set-inner-html.js            |   93.33 |    92.85 |     100 |   93.33 | 26
  describe-with-filename.js                |     100 |      100 |     100 |     100 |
  i18n-no-interpolated-values.js           |    87.5 |       80 |     100 |    87.5 | 25,31
  i18n-no-reference.js                     |   93.33 |    84.61 |     100 |     100 | 6,18,53,58
  i18n-no-tagged-templates.js              |   86.66 |    76.92 |     100 |   86.66 | 27,33
  i18n-no-template-literal.js              |   91.66 |    82.35 |     100 |   91.66 | 27,33
  no-sinon-assert-called-if-called-with.js |   92.85 |    92.85 |     100 |   92.59 | 23,29
  only-log-strings.js                      |   94.11 |    95.23 |     100 |   94.11 | 79
  only-tsx-files.js                        |     100 |      100 |     100 |     100 |
  redux-app-state.js                       |     100 |      100 |     100 |     100 |
  sort-destructured-props.js               |     100 |      100 |     100 |     100 |
  with-router-hoc-first.js                 |   94.11 |    93.75 |     100 |   93.75 | 24
-------------------------------------------|---------|----------|---------|---------|-------------------
```

File: lib/rules/i18n-no-reference.js (hunk 12 above)
Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:41]: It's actually going to pass the AST type, not the literal argument. I could look into seeing if I can get a stringified version of the argument though.

Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:41]: Got it working with the literal. Added to both rules and cleaned up the messaging a bit.

File: lib/rules/i18n-no-reference.js (hunk 11 above)
Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:35]: Done.

File: lib/rules/i18n-no-reference.js [hunk 13]
```diff
@@ -0,0 +1,93 @@
+'use strict';
+
+const { I18N_METHODS } = require('../i18n.js');
+
+function isReferenceNode(node) {
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:5]: Done.

File: lib/rules/i18n-no-reference.js (hunk 2 above)
Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:52]: Done.

File: lib/rules/i18n-no-reference.js (hunk 3 above)
Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:54]: Done.

File: lib/rules/i18n-no-reference.js (hunk 4 above)
Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:70]: Looking at it again, it actually makes more sense to add sprintf with positions [0] to the i18n config. It is totally valid for the other rules and means we don't even need this conditional at all. wdyt?

File: lib/rules/i18n-no-reference.js (hunk 5 above)
Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:75]: done.

File: lib/rules/i18n-no-template-literal.js (hunk 6 above)
Comment (KevinMind) [reply lib/rules/i18n-no-template-literal.js:32]: Done.

File: lib/rules/i18n-no-template-literal.js (hunk 7 above)
Comment (KevinMind) [reply lib/rules/i18n-no-template-literal.js:36]: Done.

File: lib/rules/i18n-no-template-literal.js (hunk 8 above)
Comment (KevinMind) [reply lib/rules/i18n-no-template-literal.js:41]: Done.

File: tests/utils.js (hunk 10 above)
Comment (KevinMind) [reply tests/utils.js:1]: Essentially yes, it is somewhat implicit.

I could add test coverage but it would just be validating that the transformation of inputs to outputs is waht the test runner already expects. Only way I could really be sure of that is by then passing the outputs into the test runner, which we already do by just using the function.

Only benefit I could see of covering this code with tests is if we see a known bug which from looking at the outputs, it is behaving as I expect it to.

Wdyty? Should I try to add tests here? Do you know what you would want to test?

**Review by KevinMind — COMMENTED** [review]
Responses

File: lib/rules/i18n-no-reference.js (hunk 4 above)
Comment (willdurand) [reply lib/rules/i18n-no-reference.js]: sounds good

File: README.md [hunk 14]
```diff
@@ -63,11 +63,12 @@ You can use the `typescript` preset to get reasonable defaults (it includes the
 <!-- THIS SECTION IS AUTOMATICALLY GENERATED, PLEASE RUN: `npm run build-doc` -->
 
 <!--DOC_START-->
-
 - [`dangerously-set-inner-html`](#dangerously-set-inner-html)
```
Comment (willdurand) [inline README.md:66]: Sorry, you'll need to rebase. I updated the README file after I saw these unrelated changes here.

File: lib/rules/i18n-no-reference.js [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+module.exports = {
+  meta: {
+    docs: {
+      githubIssue: 'https://github.com/mozilla/eslint-plugin-amo/issues/232',
+      category: 'Possible Errors',
+      description:
+        'Ensure predictable static values are passed as i18n method arguments',
+      recommended: true,
+    },
+    fixable: null, // Indicates that some problems reported by this rule are automatically fixable
```
Comment (willdurand) [inline lib/rules/i18n-no-reference.js:38]: I don't think this comment is accurate. WDYT?

**Review by willdurand — APPROVED** [review]
Looks good, thanks.  I am not sold on the `i18n-no-template-literal` rule but I am not going to block this PR for that.

r+wc

File: lib/rules/i18n-no-reference.js (hunk 15 above)
Comment (KevinMind) [reply lib/rules/i18n-no-reference.js:38]: Fixed.

File: README.md (hunk 14 above)
Comment (KevinMind) [reply README.md:66]: Rebased. Didn't change.

**Review by willdurand — APPROVED** [review]
